e3db/identity.py

```diff
@@ -45,8 +122,84 @@ def derive_note_creds(user_name: str, password: str, realm_name: str):
         password_bytes = Crypto.to_bytes(password)
         name_seed_bytes = Crypto.to_bytes(name_seed)
         key_pair = Crypto.derive_crypto_keypair(password_bytes, name_seed_bytes)
-
         salt_in_bytes = Crypto.to_bytes(key_pair.public_key + key_pair.private_key) 
         signing_key_pair = Crypto.derive_signing_keypair(password_bytes, salt_in_bytes)
 
         return note_name_as_b64_string, key_pair, signing_key_pair
+
+### Helper functions for this module ####
+
+def __response_check(response):
+    """
+    Raises errors based on response HTTP status code.
+
+    Parameters
+    ----------
+    response : requests.models.Response
+        Response object to get status code from.
+
+    Returns
+    -------
+    None
+    """
+
+    # Map of HTTP error codes to exception messages
+    # 401 Unauthorized Maps to Invalid Credentials
+    errors = {
+        400: APIError('Invalid request: HTTP 400'),
+        401: APIError('Invalid credentials: HTTP 401'),
+        403: APIError('Unauthorized: HTTP 403'),
+        404: APIError('Requested item not found: HTTP 404'),
+        429: APIError('Too many requests made to endpoint. HTTP 429. Slow down.')
+    }
+
+    # Lookup type of error we should throw, and do so if needed.
+    if response.status_code in errors:
+        raise errors[response.status_code]
+
+    # If we do not have a pre-formulated error to return, but get another HTTP
+    # Error, we check if response is in the 4XX-5XX Range, and return a generic HTTP error
+    if response.status_code >= 400 and response.status_code <= 600:
+        raise APIError("HTTP Error: {0}".format(response.status_code))
+
+def pkce_submit_challenge(user_name, realm_name, app_name, pkce_challenge, auth, api_url):
```

Comment:
Can we name this initiate login flow, since this contains a pkce verifier but also does not require it

Reply:
I would argue that it is required for this login flow to work as we want. However, I agree that removing the pkce from the method names and making them more descriptive to their outcomes is not a bad idea

Reply:
OK, I am brining my own bias here (I have only ever used an implicit flow) but I wanted to make it more clear what the code is doing. I like having explicit reference to the fact that we are utilizing a PKCE login flow - its helps explain why there are so many calls here. I tried to logically group each external call and identify why each call is necessary. The purpose of the call inside this function is to submit the verifier. Ok so now I want to change it to pkce_submit_verifier. If you disagree with this logic, please suggest an alternative- I will not die on this hill - and bc naming is hard 😬

Reply:
Since we have chosen to implement PKCE login flow here, it makes sense to me to use the verbiage of PKCE as a way to explain what the code does. To me initiate login flow is less descriptive.

Reply:
My only issue with that is that its a login endpoint but if we ever decide to expand to other login flows, we would just need to make a new method instead of expanding this one.